src/backend/distributed/executor/adaptive_executor.c

```diff
@@ -1858,10 +1843,21 @@ AssignTasksToConnectionsOrWorkerPool(DistributedExecution *execution)
 				UpdateConnectionWaitFlags(session,
 										  WL_SOCKET_READABLE | WL_SOCKET_WRITEABLE);
 
-				/* If the connections are already avaliable, make sure to activate
+				/*
+				 * If the connections are already avaliable, make sure to activate
 				 * 2PC when necessary.
 				 */
 				Activate2PCIfModifyingTransactionExpandsToNewNode(session);
+
+				/*
+				 * The previous command in the same transaction has accessed to shards
+				 * in parallel (e.g., more than 1 connection). When we are at the second
+				 * connection, we mark the relations as accessed in parallel.
+				 */
+				if (list_length(workerPool->sessionList) == 2)
+				{
+					RecordParallelRelationAccessForTaskList(execution->tasksToExecute);
```

Comment:
I considered to add the following block right after `RecordParallelRelationAccessForTaskList `, which makes sense and expected. 

```C
					/*
					 * Without use explicitly setting the execution mode to sequential,
					 * this case is unexpected. First, if citus.multi_shard_modify_mode
					 * is sequential (set by user or by the executor), we wouldn't need
					 * more than one connection. Second, the executor would have already
					 * errored out if this is a parallel access that would cause deadlocks
					 * or visibility issues due to foreign keys to reference tables.
					 *
					 * Still, the user might decide to switch to sequential execution
					 * in the middle of a transaction, and lets be prepared for it.
					 */
					if (MultiShardConnectionType == SEQUENTIAL_CONNECTION)
					{
						ereport(ERROR, (errmsg("cannot execute parallel query in "
											   "sequential execution mode"),
										errhint("Try re-running the transaction with "
												"\"SET LOCAL citus.multi_shard_modify_mode TO "
												"\'sequential\';\"")));
```


It seemed to hit two things. First, we have some tests, where we  set sequential mode in the middle of a tx block, where it actually makes sense to error.  Second, which is a lot more  important, the multi-row INSERT logic manually marks  sequential mode, so  tx blocks like the following fails:

```SQL
BEGIN;

COPY to a table which opens many connections

multi-row INSERT (which may use multiple connection in the overall distributed execution, which hits this check)  

```
So, I decided to tackle the weird  approach in m-row inserts later to keep this PR simple enough.